easypost/resource.py

```diff
@@ -0,0 +1,284 @@
+
+        try:
+            result = requests_session.request(
+                method,
+                abs_url,
+                headers=headers,
+                data=data,
+                timeout=timeout,
+                verify=True,
+            )
+            http_body = result.text
+            http_status = result.status_code
+        except Exception as e:
+            raise Error(
+                "Unexpected error communicating with EasyPost. If this "
+                f"problem persists please let us know at {SUPPORT_EMAIL}.",
+                original_exception=e,
+            )
+        return http_body, http_status
+
+    def urlfetch_request(self, method, abs_url, headers, params):
+        from easypost import SUPPORT_EMAIL, timeout
+
+        fetch_args = {"method": method, "headers": headers, "validate_certificate": False, "deadline": timeout}
+        if method == "post" or method == "put":
+            fetch_args["url"] = abs_url
+            fetch_args["payload"] = json.dumps(params, default=_utf8)
+        elif method == "get" or method == "delete":
+            fetch_args["url"] = add_params_to_url(abs_url, params)
+        else:
+            raise Error(f"Bug discovered: invalid request method: {method}. Please report to {SUPPORT_EMAIL}.")
+
+        try:
+            from google.appengine.api import urlfetch
+
+            result = urlfetch.fetch(**fetch_args)
+        except Exception as e:
+            raise Error(
+                "Unexpected error communicating with EasyPost. "
+                f"If this problem persists, let us know at {SUPPORT_EMAIL}.",
+                original_exception=e,
+            )
+
+        return result.content, result.status_code
+
+    def handle_api_error(self, http_status, http_body, response):
+        try:
+            error = response["error"]
+        except (KeyError, TypeError):
+            raise Error("Invalid response from API: (%d) %r " % (http_status, http_body), http_status, http_body)
+
+        try:
+            raise Error(error.get("message", ""), http_status, http_body)
+        except AttributeError:
+            raise Error(error, http_status, http_body)
+
+
+def _utf8(value):
+    # Python3's str(bytestring) returns "b'bytestring'" so we use .decode instead
+    if isinstance(value, bytes):
+        return value.decode("utf-8")
+    return value
+
+
+def encode_url_params(params):
+    converted_params = []
+    for key, value in sorted(params.items()):
+        if value is None:
+            continue  # don't add Nones to the query
+        elif isinstance(value, datetime.datetime):
+            value = int(time.mktime(value.timetuple()))  # to UTC timestamp
+        converted_params.append((key, value))
+        return urlencode(converted_params)
+
+
+def add_params_to_url(url, params):
+    encoded_params = encode_url_params(params)
+    if encoded_params:
+        return "%s?%s" % (url, encoded_params)
+    return url
```

Comment:
These 3 functions should probably live inside the `Requestor` class since that's the only class that uses them.